### PR TITLE
Timeouts for all requests

### PR DESCRIPTION
**Pull request by dvdplm**
Variant of #367

This PR takes a more opinionated stance than #367, where timeouts are optional. In this PR I suggest we make a all requests use a timeout and only let users choose the length.

### CODE REVIEW

File: ws-client/src/client.rs [hunk 1]
```diff
@@ -313,7 +315,16 @@ impl Client for WsClient {
...
+
+		let mut sender = self.to_back.clone();
+		let fut = sender.send(FrontToBack::Notification(raw));
+
+		let timeout = crate::tokio::sleep(self.request_timeout);
+		futures::pin_mut!(fut, timeout);
+		let res = match futures::future::select(fut, timeout).await {
+			futures::future::Either::Left((res, _)) => res,
+			futures::future::Either::Right((_, _)) => return Err(Error::RequestTimeout),
+		};
```
Comment (maciejhirsz) [inline ws-client/src/client.rs:327]: I reckon using `futures::select!` would improve readability here.

File: ws-client/src/helpers.rs [hunk 2]
```diff
@@ -188,3 +189,16 @@ pub fn process_error_response(manager: &mut RequestManager, err: JsonRpcError) -
...
 	}
 }
+
+/// Wait for a stream to complete within the given timeout.
+pub async fn call_with_timeout<T>(
+	timeout: Duration,
+	rx: oneshot::Receiver<Result<T, Error>>,
+) -> Result<Result<T, Error>, oneshot::Canceled> {
+	let timeout = crate::tokio::sleep(timeout);
+	futures::pin_mut!(rx, timeout);
```
Comment (maciejhirsz) [inline ws-client/src/helpers.rs:199]: ditto about `select!` macro

**Review by maciejhirsz — APPROVED** [review]
Looks good, minor grumble.

File: ws-client/Cargo.toml [hunk 3]
```diff
@@ -22,7 +22,7 @@ tokioV02-util = { package="tokio-util", version = "0.3", features = ["compat"],
 
 async-trait = "0.1"
 fnv = "1"
-futures = { version = "0.3.14", default-features = false, features = ["std"] }
+futures = { version = "0.3.14", default-features = false, features = ["std", "async-await"] }
```
Comment (dvdplm) [inline ws-client/Cargo.toml:25]: @maciejhirsz I wonder if this was the reason for not using `select!`, to keep the required features down?

Comment (maciejhirsz) [reply ws-client/Cargo.toml:25]: Interesting. The `select!` in futures also seems to have some weird requirements. Tokio has a `select!` macro that works identical, but doesn't require any `.fuse()` or `pin_mut!()` calls.

**Pull request merged**